client.py

```diff
@@ -161,97 +200,92 @@ def close(self):
                 asyncio.cancel(_keepalive)
-                await asyncio.sleep(1)  # wait for cancellation
                 if self._concb is not None:
                     # apps might need to know if they lost connection to the server
                     launch(self._concb, False, *self._concbargs)
-            finally:
-                init = False
-                self.close()  # Close socket
-                s.disconnect()
-                await asyncio.sleep_ms(self._to * 2)  # Ensure server detects outage
-                while s.isconnected():
-                    await asyncio.sleep(1)
+            else:
+                asyncio.cancel(_reader)
+            self._verbose and print(self._evfail.value())
+            await asyncio.sleep(1)  # wait for cancellation but notify application before
+            init = False
+            self.close()  # Close socket
+            s.disconnect()
+            # await asyncio.sleep_ms(self._to * 2)  # Ensure server detects outage.
+            # server should detect outage if client reconnects or no ACKs are received in qos
+            while s.isconnected():
+                await asyncio.sleep(1)
 
     async def _reader(self):  # Entry point is after a (re) connect.
         c = self.connects  # Count successful connects
         self._evread.clear()  # No data read yet
         try:
             while True:
-                line = await self._readline()  # OSError on fail
-                mid = int(line[0:2], 16)
-                if len(line) == 3:  # Got ACK: remove from expected list
-                    self._acks_pend.discard(mid)  # qos0 acks are ignored
-                    continue  # All done
-                # Old message still pending. Discard new one peer will re-send.
-                if self._evread.is_set():
+                preheader, header, line = await self._readline()  # OSError on fail
+                print("Got", preheader, header, line)
+                # Discard dupes
+                mid = preheader[0]
+                if preheader[4] == 0x2C:  # ACK
+                    print("Got ack mid", mid)
+                    self._ack_mid = mid
                     continue
-                # Message received & can be passed to user: send ack.
-                self._loop.create_task(self._sendack(mid))
-                # Discard dupes. mid == 0 : Server has power cycled
-                if not mid:
-                    isnew(-1)  # Clear down rx message record
-                if isnew(mid):
-                    self._evread.set(line[2:].decode())
+                if mid != self._recv_mid:
+                    # Read succeeded: flag .readline
+                    if self._evread.is_set():
+                        self._verbose and print("Dumping unread message", self._evread.value())
+                    self._evread.set((header, line))
+                    self._recv_mid = mid
+                if preheader[4] & 0x01 == 1:  # qos==True, send ACK even if dupe
```

Comment:
Only sending an ACK back for qos messages as we can distinguish between qos and non-qos messages.